typing/parmatch.ml

```diff
@@ -1483,61 +1476,53 @@ let rec exhaust (ext:Path.t option) pss n = match pss with
-          | Witnesses r ->
-              let q0 = Pattern_head.to_omega_pattern q0 in
-              Witnesses (List.map (fun row -> q0::row) r)
-          | r -> r
-        end
+          let sub_witnesses = exhaust ext default (n-1) in
+          let q0 = Pattern_head.to_omega_pattern q0 in
+          Seq.map (fun row -> q0::row) sub_witnesses
       | { default; constrs } ->
           let try_non_omega (p,pss) =
             if is_absent_pat p then
-              No_matching_value
+              Seq.empty
             else
-              match
+              let sub_witnesses =
                 exhaust
                   ext pss
                   (List.length (simple_match_args p Pattern_head.omega [])
                    + n - 1)
-              with
-              | Witnesses r ->
-                  let p = Pattern_head.to_omega_pattern p in
-                  Witnesses (List.map (set_args p) r)
-              | r       -> r in
-          let before = try_many try_non_omega constrs in
-          if
-            full_match false constrs && not (should_extend ext constrs)
-          then
-            before
-          else
-            let r =  exhaust ext default (n-1) in
-            match r with
-            | No_matching_value -> before
-            | Witnesses r ->
-                try
-                  let p = build_other ext constrs in
-                  let dug = List.map (fun tail -> p :: tail) r in
-                  match before with
-                  | No_matching_value -> Witnesses dug
-                  | Witnesses x -> Witnesses (x @ dug)
-                with
-        (* cannot occur, since constructors don't make a full signature *)
-                | Empty -> fatal_error "Parmatch.exhaust"
-  end
+              in
+              let p = Pattern_head.to_omega_pattern p in
+              Seq.map (set_args p) sub_witnesses
+          in
+          let try_omega () =
+            if full_match false constrs && not (should_extend ext constrs) then
+              Seq.empty
+            else
+              let sub_witnesses = exhaust ext default (n-1) in
+              match build_other ext constrs with
+              | exception Empty ->
+                  (* cannot occur, since constructors don't make
+                     a full signature *)
+                  fatal_error "Parmatch.exhaust"
+              | p ->
+                  Seq.map (fun tail -> p :: tail) sub_witnesses
+          in
+          (* Lazily compute witnesses for all constructor submatrices
+             (Some constr_mat) then the wildcard submatrix (None).
+             Note that the call to [try_omega ()] is delayed to after
+             all constructor matrices have been traversed. *)
+          List.map (fun constr_mat -> Some constr_mat) constrs @ [None]
+          |> List.to_seq
+          |> Seq.flat_map
+            (function
+              | Some constr_mat -> try_non_omega constr_mat
+              | None -> try_omega ())
```

Comment:
I found this a bit hard to get my head around, harder than I'd like it to be at least.
The following seems simpler to me
```ocaml
Seq.append
  (Seq.flat_map try_non_omega (List.to_seq constrs))
  (try_omega ())
```
Though I understand it implies more computations early on (but does it really matter?)

Note: I'm fine with leaving the code as is.

Reply:
I care about delaying computations as much as we (easily) can. Should I use ``[ `Specialized of foo | `Default ]`` instead of options?